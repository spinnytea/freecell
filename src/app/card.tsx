import Image from "next/image";

// TODO red & black joker
export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];

export type Rank = 'ace' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | 'jack' | 'queen' | 'king';
export const RankList: Rank[] = ['ace', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'jack', 'queen', 'king'];

const ORIG_WIDTH = 167.0869141;
const ORIG_HEIGHT = 242.6669922;
const scale_height = (width: number) => Math.floor(width / ORIG_WIDTH * ORIG_HEIGHT);

export function CardImage({
	width=ORIG_WIDTH,
	rank,
	suit,
}: {
	width?: number,
	rank: Rank,
	suit: Suit,
}) {
	const filename = getFilename(rank, suit);
	const height = scale_height(width);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={width}
			height={height}
			priority
		/>
	);
}

function getFilename(rank: Rank, suit: Suit) {
	const fancy = (rank === 'jack' || rank === 'queen' || rank === 'king' ? '2' : '');
	return `/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}
