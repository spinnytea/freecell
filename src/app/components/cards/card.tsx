import Image from 'next/image';

// TODO red & black joker
export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];

const ORIG_WIDTH = 167.0869141;
const ORIG_HEIGHT = 242.6669922;
const scale_height = (width: number) => Math.floor((width / ORIG_WIDTH) * ORIG_HEIGHT);

const FANCY_DECK = true;

export function CardImage({
	width = ORIG_WIDTH,
	rank,
	suit,
}: {
	width?: number;
	rank: Rank;
	suit: Suit;
}) {
	// REVIEW do we need `priority` - it's complaining because i'm rendering the whole deck on init
	const filename = getFilename(rank, suit);
	const height = scale_height(width);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			priority
		/>
	);
}

// TODO toggle fancy (also ace of spades)
function getFilename(rank: Rank, suit: Suit, useFancyDeck = FANCY_DECK) {
	if (rank === 'joker') {
		if (isRed(suit)) return `/SVG-cards-1.3/red_joker.svg`;
		return `/SVG-cards-1.3/black_joker.svg`;
	}
	if (rank === 'ace' && suit === 'spades') {
		if (useFancyDeck) return `/SVG-cards-1.3/ace_of_spades.svg`;
		return `/SVG-cards-1.3/ace_of_spades2.svg`;
	}
	const fancy = useFancyDeck
		? rank === 'jack' || rank === 'queen' || rank === 'king'
			? '2'
			: ''
		: '';
	return `/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}

/** stuff exported for testing, not meant to be used directly */
export const units = {
	getFilename,
};
