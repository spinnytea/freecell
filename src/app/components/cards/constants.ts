export const ORIG_WIDTH = 208;
export const ORIG_HEIGHT = 303;

export const ASSET_FOLDER = process.env.BASE_PATH ?? '';

export type CardFaces = 'SVGCards13' | 'SmolCards';
