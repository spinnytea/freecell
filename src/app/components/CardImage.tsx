import Image from 'next/image';
import { isRed, Rank, Suit } from '@/app/game/card';

const ORIG_WIDTH = 167.0869141;
const ORIG_HEIGHT = 242.6669922;
export const scale_height = (width: number) => Math.floor((width / ORIG_WIDTH) * ORIG_HEIGHT);

const FANCY_DECK = true;
const assetFolder = process.env.BASE_PATH ?? '';

// TODO https://cardmeister.github.io/
// TODO https://en.wikipedia.org/wiki/Playing_cards_in_Unicode
export function CardImage({
	hidden = false,
	rank,
	suit,
	width = ORIG_WIDTH,
}: Readonly<{
	hidden?: boolean;
	rank: Rank;
	suit: Suit;
	width?: number;
}>) {
	// REVIEW do we need `priority` - it's complaining because i'm rendering the whole deck on init
	const filename = getFilename(rank, suit, hidden);
	const height = scale_height(width);
	return (
		<Image
			src={filename}
			alt={hidden ? 'card back' : `${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			draggable={false}
			priority
		/>
	);
}

// TODO option to toggle fancy
// TODO alternate card backs?
// TODO deck w/ kings are lions?
// TODO dark theme cards
function getFilename(rank: Rank, suit: Suit, hidden: boolean, useFancyDeck = FANCY_DECK) {
	if (hidden) {
		return `${assetFolder}/i/Card_back_10.svg`;
	}

	if (rank === 'joker') {
		if (isRed(suit)) return `${assetFolder}/i/SVG-cards-1.3/red_joker.svg`;
		return `${assetFolder}/i/SVG-cards-1.3/black_joker.svg`;
	}
	if (rank === 'ace' && suit === 'spades') {
		if (useFancyDeck) return `${assetFolder}/i/SVG-cards-1.3/ace_of_spades.svg`;
		return `${assetFolder}/i/SVG-cards-1.3/ace_of_spades2.svg`;
	}
	const fancy = useFancyDeck
		? rank === 'jack' || rank === 'queen' || rank === 'king'
			? '2'
			: ''
		: '';
	return `${assetFolder}/i/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}

/** stuff exported for testing, not meant to be used directly */
export const units = {
	getFilename,
};
