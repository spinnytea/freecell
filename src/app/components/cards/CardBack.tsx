import Image from 'next/image';
import { ASSET_FOLDER } from '@/app/components/cards/constants';

// TODO (theme) alternate card backs?
// TODO (hud) (2-priority) Airplane mode: no deck?
// - Does it check for it every time?
// - Do we need to place an invisible one / off screen so it's always loaded?
export function CardBack({
	width,
	height,
}: Readonly<{
	width: number;
	height: number;
}>) {
	const filename = getFilename();
	return (
		<Image
			src={filename}
			alt="card back"
			width={Math.floor(width)}
			height={height}
			draggable={false}
			priority
		/>
	);
}

function getFilename() {
	return `${ASSET_FOLDER}/i/Card_back_10.svg`;
}
