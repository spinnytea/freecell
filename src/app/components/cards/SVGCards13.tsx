import Image from 'next/image';
import { ASSET_FOLDER } from '@/app/components/cards/constants';
import { isRed, Rank, Suit } from '@/app/game/card';

export function SVGCards13({
	width,
	height,
	rank,
	suit,
}: Readonly<{
	width: number;
	height: number;
	rank: Rank;
	suit: Suit;
}>) {
	const filename = getFilename(rank, suit);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			draggable={false}
			priority
		/>
	);
}

function getFilename(rank: Rank, suit: Suit, useFancyDeck = true) {
	if (rank === 'joker') {
		if (isRed(suit)) return `${ASSET_FOLDER}/i/SVG-cards-1.3/red_joker.svg`;
		return `${ASSET_FOLDER}/i/SVG-cards-1.3/black_joker.svg`;
	}
	if (rank === 'ace' && suit === 'spades') {
		if (useFancyDeck) return `${ASSET_FOLDER}/i/SVG-cards-1.3/ace_of_spades.svg`;
		return `${ASSET_FOLDER}/i/SVG-cards-1.3/ace_of_spades2.svg`;
	}
	const fancy = useFancyDeck
		? rank === 'jack' || rank === 'queen' || rank === 'king'
			? '2'
			: ''
		: '';
	return `${ASSET_FOLDER}/i/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}

/** stuff exported for testing, not meant to be used directly */
export const units = {
	getFilename,
};
