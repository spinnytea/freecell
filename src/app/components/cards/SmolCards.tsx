import { ReactElement } from 'react';
import Image from 'next/image';
import { ASSET_FOLDER } from '@/app/components/cards/constants';
import { ace_of_clubs, ace_of_clubs_defs } from '@/app/components/cards/smol/ace_of_clubs';
import { ace_of_diamonds, ace_of_diamonds_defs } from '@/app/components/cards/smol/ace_of_diamonds';
import { ace_of_hearts, ace_of_hearts_defs } from '@/app/components/cards/smol/ace_of_hearts';
import { ace_of_spades, ace_of_spades_defs } from '@/app/components/cards/smol/ace_of_spades';
import { clubs } from '@/app/components/cards/smol/clubs';
import { diamonds } from '@/app/components/cards/smol/diamonds';
import { hearts } from '@/app/components/cards/smol/hearts';
import { spades } from '@/app/components/cards/smol/spades';
import { isRed, Rank, Suit } from '@/game/card/card';

const SYMBOLS: { [suit in Suit]: ReactElement } = {
	clubs,
	diamonds,
	hearts,
	spades,
};
const ACE_SYMBOL_DEFS: { [suit in Suit]: ReactElement } = {
	clubs: ace_of_clubs_defs,
	diamonds: ace_of_diamonds_defs,
	hearts: ace_of_hearts_defs,
	spades: ace_of_spades_defs,
};
const ACE_SYMBOLS: { [suit in Suit]: ReactElement } = {
	clubs: ace_of_clubs,
	diamonds: ace_of_diamonds,
	hearts: ace_of_hearts,
	spades: ace_of_spades,
};
const TEXT: { [rank in Rank]: string } = {
	'ace': 'A',
	'2': '2',
	'3': '3',
	'4': '4',
	'5': '5',
	'6': '6',
	'7': '7',
	'8': '8',
	'9': '9',
	'10': '10',
	'jack': 'J',
	'queen': 'Q',
	'king': 'K',
	'joker': 'W', // unused
};

export function SmolCards({
	width,
	height,
	rank,
	suit,
}: Readonly<{
	width: number;
	height: number;
	rank: Rank;
	suit: Suit;
}>) {
	if (rank !== 'joker') {
		return <NativeCards width={width} height={height} rank={rank} suit={suit} />;
	}

	// XXX (accessibility) consider adding title={rank or suit}
	const filename = getFilename(rank, suit);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			draggable={false}
			priority
		/>
	);
}

function NativeCards({
	width,
	height,
	rank,
	suit,
}: Readonly<{
	width: number;
	height: number;
	rank: Rank;
	suit: Suit;
}>) {
	const color = isRed(suit) ? '#df0000' : '#000000';
	const text = TEXT[rank];
	const symbol = SYMBOLS[suit];

	// XXX (accessibility) consider adding <title>rank or suit</title>
	return (
		<svg
			width={width}
			height={height}
			viewBox="0 0 208 303"
			xmlns="http://www.w3.org/2000/svg"
			xmlnsXlink="http://www.w3.org/1999/xlink"
		>
			<desc>
				{rank} of {suit}
			</desc>

			{rank === 'ace' && ACE_SYMBOL_DEFS[suit]}

			<rect width="208" height="303" rx="8" fill="#FFFFFF" stroke="#000000" strokeWidth="0.5" />

			<g fill={color} stroke="none" textAnchor="middle" dominantBaseline="middle" fontWeight={500}>
				<text fontSize="90" x="52" y="56" letterSpacing="-0.1em">
					{text}
				</text>
			</g>

			<g>{symbol}</g>

			{rank === 'ace' ? (
				<g transform="scale(1.2) translate(-19, -40)">{ACE_SYMBOLS[suit]}</g>
			) : (
				<g transform="scale(2) translate(-95, 42)">{symbol}</g>
			)}
		</svg>
	);
}

function getFilename(rank: Rank, suit: Suit) {
	if (rank === 'joker') {
		// TODO (joker) restyle joker
		//  - (make one for each suit, rotate the center so suit is top-left)
		if (isRed(suit)) return `${ASSET_FOLDER}/i/SVG-cards-1.3/red_joker.svg`;
		return `${ASSET_FOLDER}/i/SVG-cards-1.3/black_joker.svg`;
	}

	return `${ASSET_FOLDER}/i/smol-cards/ace_of_${suit}.svg`;
}
