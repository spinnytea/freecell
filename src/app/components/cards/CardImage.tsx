import { CardBack } from '@/app/components/cards/CardBack';
import { ORIG_HEIGHT, ORIG_WIDTH } from '@/app/components/cards/constants';
import { SVGCards13 } from '@/app/components/cards/SVGCards13';
import { Rank, Suit } from '@/app/game/card';

export const scale_height = (width: number) => Math.floor((width / ORIG_WIDTH) * ORIG_HEIGHT);

// TODO (theme) https://cardmeister.github.io/
//  - https://github.com/cardmeister/cardmeister.github.io/blob/master/elements.cardmeister.full.js
// TODO (theme) https://en.wikipedia.org/wiki/Playing_cards_in_Unicode
// TODO (theme) alternate card backs?
// TODO (theme) deck w/ kings are lions?
// TODO (theme) dark theme cards
// FIXME find a deck that's (tl Rank | tr Suit) (visible in stack), and the picture is just a large suit icon
//  - based on SVG-cards-1.3
// ---
//  - ace_of_clubs.svg (normal)
//  - ace_of_diamonds.svg (normal)
//  - ace_of_hearts.svg (normal)
//  - ace_of_spades.svg (fancy)
//  - ace_of_spades2.svg (normal)
// ---
//  - black_joker.svg (normal)
//  - red_joker.svg (normal)
//  - (make one for each suit, rotate the center so suit is top-left)
// ---
//  - red numbers are better than black (the 10 is weird, the rest look the same?)
export function CardImage({
	hidden = false,
	rank,
	suit,
	width = ORIG_WIDTH,
}: Readonly<{
	hidden?: boolean;
	rank: Rank;
	suit: Suit;
	width?: number;
}>) {
	const height = scale_height(width);

	if (hidden) {
		return <CardBack width={width} height={height} />;
	}

	return <SVGCards13 width={width} height={height} rank={rank} suit={suit} />;
}
