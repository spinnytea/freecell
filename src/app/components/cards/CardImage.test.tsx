import { render, screen } from '@testing-library/react';
import { CardImage, units } from '@/app/components/cards/CardImage';
import { RankList, SuitList } from '@/app/game/card';

describe('CardImage', () => {
	test('first', () => {
		render(<CardImage rank="5" suit="hearts" />);
		expect(screen.getByAltText('5 of hearts')).toMatchSnapshot();
	});

	describe('getFilename', () => {
		test('fancy', () => {
			const allNames: string[] = [];

			SuitList.forEach((suit) => {
				RankList.forEach((rank) => {
					allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit, true)}`);
				});
			});
			SuitList.forEach((suit) => {
				const rank = 'joker';
				allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit, true)}`);
			});
			// allNames.push(`card back → ${units.getFilename('ace', 'spades', true)}`);

			expect(allNames).toMatchSnapshot();
		});

		test('simple', () => {
			const allNames: string[] = [];

			SuitList.forEach((suit) => {
				RankList.forEach((rank) => {
					allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit, false)}`);
				});
			});
			SuitList.forEach((suit) => {
				const rank = 'joker';
				allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit, false)}`);
			});
			// allNames.push(`card back → ${units.getFilename('ace', 'spades', false)}`);

			expect(allNames).toMatchSnapshot();
		});
	});
});
