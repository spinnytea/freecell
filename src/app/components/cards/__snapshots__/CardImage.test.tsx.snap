// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardImage first 1`] = `
<img
  alt="5 of hearts"
  data-nimg="1"
  decoding="async"
  draggable="false"
  fetchpriority="high"
  height="303"
  src="/i/SVG-cards-1.3/5_of_hearts.svg"
  style="color: transparent;"
  width="208"
/>
`;
